Support for macOS

Hi, could you provide support for macOS?
I tried to install it in macOS but it doesn't work.

Installation shows the following

./bootstrap
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: /opt/local/bin/aclocal
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: /opt/local/bin/autoconf
autoreconf: configure.ac: not using Autoheader
autoreconf: running: /opt/local/bin/automake --add-missing --copy --no-force
autoreconf: Leaving directory '.'

./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of gcc... gcc3
checking for sin in -lm... yes
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for float.h... yes
checking for inttypes.h... (cached) yes
checking for limits.h... yes
checking for stdint.h... (cached) yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking for _Bool... yes
checking for stdbool.h that conforms to C99... yes
checking for int32_t... yes
checking for int64_t... yes
checking for size_t... yes
checking for uint32_t... yes
checking for uint64_t... yes
checking for error_at_line... no
checking build system type... aarch64-apple-darwin23.0.0
checking host system type... aarch64-apple-darwin23.0.0
checking for GNU libc compatible malloc... yes
checking for floor... yes
checking for memmove... yes
checking for memset... yes
checking for strdup... yes
checking for strstr... yes
checking for strtol... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating GNUmakefile
config.status: creating man/GNUmakefile
config.status: creating src/GNUmakefile
config.status: executing depfiles commands

make
Making all in src
depbase=`echo unified.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
	gcc -DPACKAGE_NAME=\"libpicoquant\" -DPACKAGE_TARNAME=\"libpicoquant\" -DPACKAGE_VERSION=\"0.6.2\" -DPACKAGE_STRING=\"libpicoquant\ 0.6.2\" -DPACKAGE_BUGREPORT=\"tsbischof@gmail.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libpicoquant\" -DVERSION=\"0.6.2\" -DHAVE_LIBM=1 -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_FLOAT_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_LIMITS_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE__BOOL=1 -DHAVE_MALLOC=1 -DHAVE_FLOOR=1 -DHAVE_MEMMOVE=1 -DHAVE_MEMSET=1 -DHAVE_STRDUP=1 -DHAVE_STRSTR=1 -DHAVE_STRTOL=1 -I.    -Wall -D _FILE_OFFSET_BITS=64 -g -O2 -MT unified.o -MD -MP -MF $depbase.Tpo -c -o unified.o unified.c &&\
	mv -f $depbase.Tpo $depbase.Po
unified.c:40:10: warning: the current #pragma pack alignment value is modified in the included file [-Wpragma-pack]
#include "hydraharp/hh_v10.h"
         ^
./hydraharp/hh_v10.h:38:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(2)
        ^
unified.c:134:194: error: expected ';' after expression
                                        error("This mode is not implemented due to a lack of test data. Please open a pull request at https://github.com/tsbischof/picoquant-sample-data and we will get to work on the decoder.\n")
                                                                                                                                                                                                                                    ^
                                                                                                                                                                                                                                    ;
unified.c:190:42: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
                                TAG_PRINT(fprintf(stream_out, "%ld", (uint64_t)tag.value))
                                                               ~~~   ^~~~~~~~~~~~~
~~~~~~
                                                               %llu
unified.c:50:53: note: expanded from macro 'TAG_PRINT'
#define TAG_PRINT(x) if ( options->print_header ) { x; }
                                                    ^
unified.c:194:44: warning: format specifies type 'unsigned long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
                                TAG_PRINT(fprintf(stream_out, "0x%lx", (uint64_t)tag.value))
                                                                 ~~~   ^~~~~~~~~~~~~~~~~~~
                                                                 %llx
unified.c:50:53: note: expanded from macro 'TAG_PRINT'
#define TAG_PRINT(x) if ( options->print_header ) { x; }
                                                    ^
unified.c:206:47: warning: format specifies type 'unsigned long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
                                TAG_PRINT(fprintf(stream_out, "0x%016lx", tag.value))
                                                                 ~~~~~~   ^~~~~~~~~
                                                                 %016llx
unified.c:50:53: note: expanded from macro 'TAG_PRINT'
#define TAG_PRINT(x) if ( options->print_header ) { x; }
                                                    ^
4 warnings and 1 error generated.
make[1]: *** [unified.o] Error 1
make: *** [all-recursive] Error 1
➜  libpicoquant git:(master)
➜  libpicoquant git:(master) ls
GNUmakefile    TODO           config.guess   depcomp        sample_data
GNUmakefile.am VERSION        config.log     doc            src
GNUmakefile.in aclocal.m4     config.status  install-sh
INSTALL        autom4te.cache config.sub     man
LICENSE        bootstrap      configure      missing
README.md      compile        configure.ac   python
➜  libpicoquant git:(master) vim INSTALL
➜  libpicoquant git:(master) ls
GNUmakefile    TODO           config.guess   depcomp        sample_data
GNUmakefile.am VERSION        config.log     doc            src
GNUmakefile.in aclocal.m4     config.status  install-sh
INSTALL        autom4te.cache config.sub     man
LICENSE        bootstrap      configure      missing
README.md      compile        configure.ac   python
➜  libpicoquant git:(master) picoquant --file-in "data.phd"
zsh: command not found: picoquant
➜  libpicoquant git:(master) ./bootstrap
autoreconf: export WARNINGS=
autoreconf: Entering directory '.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: /opt/local/bin/aclocal
autoreconf: configure.ac: tracing
autoreconf: configure.ac: not using Libtool
autoreconf: configure.ac: not using Intltool
autoreconf: configure.ac: not using Gtkdoc
autoreconf: running: /opt/local/bin/autoconf
autoreconf: configure.ac: not using Autoheader
autoreconf: running: /opt/local/bin/automake --add-missing --copy --no-force
autoreconf: Leaving directory '.'
➜  libpicoquant git:(master) ./configure
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a race-free mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... no
checking for awk... awk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking whether gcc understands -c and -o together... yes
checking whether make supports the include directive... yes (GNU style)
checking dependency style of gcc... gcc3
checking for sin in -lm... yes
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for float.h... yes
checking for inttypes.h... (cached) yes
checking for limits.h... yes
checking for stdint.h... (cached) yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking for _Bool... yes
checking for stdbool.h that conforms to C99... yes
checking for int32_t... yes
checking for int64_t... yes
checking for size_t... yes
checking for uint32_t... yes
checking for uint64_t... yes
checking for error_at_line... no
checking build system type... aarch64-apple-darwin23.0.0
checking host system type... aarch64-apple-darwin23.0.0
checking for GNU libc compatible malloc... yes
checking for floor... yes
checking for memmove... yes
checking for memset... yes
checking for strdup... yes
checking for strstr... yes
checking for strtol... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating GNUmakefile
config.status: creating man/GNUmakefile
config.status: creating src/GNUmakefile
config.status: executing depfiles commands
➜  libpicoquant git:(master) make
Making all in src
depbase=`echo unified.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
	gcc -DPACKAGE_NAME=\"libpicoquant\" -DPACKAGE_TARNAME=\"libpicoquant\" -DPACKAGE_VERSION=\"0.6.2\" -DPACKAGE_STRING=\"libpicoquant\ 0.6.2\" -DPACKAGE_BUGREPORT=\"tsbischof@gmail.com\" -DPACKAGE_URL=\"\" -DPACKAGE=\"libpicoquant\" -DVERSION=\"0.6.2\" -DHAVE_LIBM=1 -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DSTDC_HEADERS=1 -DHAVE_FLOAT_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_LIMITS_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE__BOOL=1 -DHAVE_MALLOC=1 -DHAVE_FLOOR=1 -DHAVE_MEMMOVE=1 -DHAVE_MEMSET=1 -DHAVE_STRDUP=1 -DHAVE_STRSTR=1 -DHAVE_STRTOL=1 -I.    -Wall -D _FILE_OFFSET_BITS=64 -g -O2 -MT unified.o -MD -MP -MF $depbase.Tpo -c -o unified.o unified.c &&\
	mv -f $depbase.Tpo $depbase.Po
unified.c:40:10: warning: the current #pragma pack alignment value is modified in the included file [-Wpragma-pack]
#include "hydraharp/hh_v10.h"
         ^
./hydraharp/hh_v10.h:38:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(2)
        ^
unified.c:134:194: error: expected ';' after expression
                                        error("This mode is not implemented due to a lack of test data. Please open a pull request at https://github.com/tsbischof/picoquant-sample-data and we will get to work on the decoder.\n")
                                                                                                                                                                                                                                    ^
                                                                                                                                                                                                                                    ;
unified.c:190:42: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
                                TAG_PRINT(fprintf(stream_out, "%ld", (uint64_t)tag.value))
                                                               ~~~   ^~~~~~~~~~~~~~~~~~~
                                                               %llu
unified.c:50:53: note: expanded from macro 'TAG_PRINT'
#define TAG_PRINT(x) if ( options->print_header ) { x; }
                                                    ^
unified.c:194:44: warning: format specifies type 'unsigned long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
                                TAG_PRINT(fprintf(stream_out, "0x%lx", (uint64_t)tag.value))
                                                                 ~~~   ^~~~~~~~~~~~~~~~~~~
                                                                 %llx
unified.c:50:53: note: expanded from macro 'TAG_PRINT'
#define TAG_PRINT(x) if ( options->print_header ) { x; }
                                                    ^
unified.c:206:47: warning: format specifies type 'unsigned long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
                                TAG_PRINT(fprintf(stream_out, "0x%016lx", tag.value))
                                                                 ~~~~~~   ^~~~~~~~~
                                                                 %016llx
unified.c:50:53: note: expanded from macro 'TAG_PRINT'
#define TAG_PRINT(x) if ( options->print_header ) { x; }
                                                    ^
4 warnings and 1 error generated.
make[1]: *** [unified.o] Error 1
make: *** [all-recursive] Error 1

picoquant
zsh: command not found: picoquant

I have gcc, clang installed in my mac.

--- a/.gitignore
+++ b/.gitignore
@@ -184,3 +184,7 @@ modules.order
 Module.symvers
 Mkfile.old
 dkms.conf
+
+*.log
+*.trs
+*~

--- a/GNUmakefile.am
+++ b/GNUmakefile.am
@@ -1,2 +1,3 @@
 AUTOMAKE_OPTIONS = foreign subdir-objects
 SUBDIRS = src man
+TESTS = ./test.py

--- a/README.md
+++ b/README.md
@@ -3,13 +3,14 @@
 It is intended to provide a generic interface to these formats which is compatible with other analysis code (e.g. [photon_correlation](github.com/tsbischof/photon_correlation.git)). 
 
 ## Building
-This program is tested primarily on Linux, both 32-bit and 64-bit.
-It has no external dependencies apart from a C compiler (tested with clang and gcc).
+This program is tested primarily on Linux (64-bit).
+It has no external dependencies apart from a C compiler for build (tested mostly with gcc) and python3 for test.
 To build:
 ```
 ./bootstrap
 ./configure
-make
+make check
+make install
 ```
 
 ## Usage

--- a/VERSION
+++ b/VERSION
@@ -1,3 +1,11 @@
+0.6.3 (2023-11-11)
+
+* add unit tests
+* rework #pragma pack calls to suppress warnings on clang version>=6
+* add missing semicolon which was breaking the build
+
+--------------------------------------------------------------------------------
+
 0.6.2 (2023-03-29)
 
 * bugfix: ptu hh v2 t3 decoder had the same bug as the v1, thank you to Adam Olejniczak for the bug report and sample data!

--- a/configure.ac
+++ b/configure.ac
@@ -1,5 +1,5 @@
 AC_PREREQ([2.69])
-AC_INIT(libpicoquant, 0.6.2, tsbischof@gmail.com)
+AC_INIT(libpicoquant, 0.6.3, tsbischof@gmail.com)
 AC_CONFIG_SRCDIR([src])
 AM_INIT_AUTOMAKE
 
@@ -26,4 +26,5 @@ AC_FUNC_MALLOC
 AC_CHECK_FUNCS([floor memmove memset strdup strstr strtol])
 
 AC_CONFIG_FILES([GNUmakefile man/GNUmakefile src/GNUmakefile])
+
 AC_OUTPUT

--- a/sample_data/generate_contents
+++ b/sample_data/generate_contents
@@ -7,6 +7,7 @@ import subprocess
 
 is_picoquant_file = re.compile(".+\.(p[ht]u|[hp]t[d23]|[hpt]hd|t3r)$")
 
+
 def _process(cmd, src, suffix, force, n_lines=None):
     dst = src + suffix
     if force or not os.path.exists(dst):
@@ -17,50 +18,80 @@ def _process(cmd, src, suffix, force, n_lines=None):
         else:
             full_cmd = ["picoquant", "--file-in", src] + cmd
             records = subprocess.Popen(full_cmd, stdout=subprocess.PIPE)
-            truncated = subprocess.Popen(["head", "--lines={}".format(n_lines)],
-                                         stdin=records.stdout,
-                                         stdout=subprocess.PIPE)
+            truncated = subprocess.Popen(
+                ["head", "--lines={}".format(n_lines)],
+                stdin=records.stdout,
+                stdout=subprocess.PIPE,
+            )
             with open(dst, "wb") as stream_out:
                 stream_out.write(truncated.stdout.read())
+
 
 def get_data(filename, force=False, n_lines=None):
     _process([], filename, ".data", force, n_lines=n_lines)
 
+
 def get_header(filename, force=False):
     _process(["--header-only"], filename, ".header", force)
+
 
 def get_resolution(filename, force=False):
     _process(["--resolution-only"], filename, ".resolution", force)
 
+
 def process_files(root_dir, n_data_lines=None, force=False):
     for root, dirs, files in os.walk(root_dir):
-        for filename in filter(lambda f: is_picoquant_file.match(f) and \
-                                         os.path.isfile(os.path.join(root, f)),
-                               files):
+        for filename in filter(
+            lambda f: is_picoquant_file.match(f)
+            and os.path.isfile(os.path.join(root, f)),
+            files,
+        ):
             filename = os.path.join(root, filename)
             get_data(filename, force=force, n_lines=n_data_lines)
             get_header(filename, force=force)
             get_resolution(filename, force=force)
 
+
 def clean_output(root_dir):
     for root, dirs, files in os.walk(root_dir):
         for filename in filter(
-              lambda _: re.search("\.(data|header|resolution)$", _), 
-              files):
+            lambda _: re.search("\.(data|header|resolution)$", _), files
+        ):
             os.remove(os.path.join(root, filename))
+
 
 def main():
     parser = argparse.ArgumentParser()
-    parser.add_argument("--clean", dest="clean", action="store_true", 
-                        default=False, 
-                        help="Remove output files like *.data, *.header, and *.resolution")
-    parser.add_argument("--force", dest="force", action="store_true", 
-                        default=False,
-                        help="If output files exist, overwrite them. Otherwise, the file will not be regenerated.")
-    parser.add_argument("--lines", dest="lines", type=int, default=None, 
-                        help="Number of lines of data to include. By default, all are included.")
-    parser.add_argument("--dir", dest="root_dir", type=str, default=os.getcwd(),
-                        help="Process files in the specified location. By default, the current directory is used.")
+    parser.add_argument(
+        "--clean",
+        dest="clean",
+        action="store_true",
+        default=False,
+        help="Remove output files like *.data, *.header, and *.resolution",
+    )
+    parser.add_argument(
+        "--force",
+        dest="force",
+        action="store_true",
+        default=False,
+        help="If output files exist, overwrite them. Otherwise, the "
+        "file will not be regenerated.",
+    )
+    parser.add_argument(
+        "--lines",
+        dest="lines",
+        type=int,
+        default=None,
+        help="Number of lines of data to include. By default, all are included.",
+    )
+    parser.add_argument(
+        "--dir",
+        dest="root_dir",
+        type=str,
+        default=os.getcwd(),
+        help="Process files in the specified location. By default, "
+        "the current directory is used.",
+    )
 
     args = parser.parse_args()
 
@@ -68,6 +99,7 @@ def main():
         clean_output(args.root_dir)
     else:
         process_files(args.root_dir, n_data_lines=args.lines, force=args.force)
+
 
 if __name__ == "__main__":
     main()

--- a/src/header.h
+++ b/src/header.h
@@ -32,10 +32,9 @@
 #ifndef HEADER_H_
 #define HEADER_H_
 
-#include <stdio.h>
+#pragma pack(push, 2)
 
-// Enforce byte alignment to ensure cross-platform compatibility.
-//#pragma pack(4)
+#include <stdio.h>
 
 typedef struct {
 	char Ident[8];
@@ -51,5 +50,7 @@ int pq_unified_header_read(FILE *in_stream, pq_header_t *pq_header, pu_header_t 
 int pq_header_read(FILE *in_stream, pq_header_t *pq_header);
 void pq_header_printf(FILE *out_stream, pq_header_t *pq_header);
 void pq_header_fwrite(FILE *out_stream, pq_header_t *pq_header);
+
+#pragma pack(pop)
 
 #endif 

--- a/src/hydraharp/hh_v10.h
+++ b/src/hydraharp/hh_v10.h
@@ -32,10 +32,11 @@
 #ifndef HH_V10_H_
 #define HH_V10_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 #include "../picoquant.h"
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -245,5 +246,7 @@ void hh_v10_tttr_header_printf(FILE *stream_out,
 		hh_v10_tttr_header_t *tttr_header);
 void hh_v10_tttr_header_fwrite(FILE *stream_out, 
 		hh_v10_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/hydraharp/hh_v20.h
+++ b/src/hydraharp/hh_v20.h
@@ -32,10 +32,11 @@
 #ifndef HH_V20_H_
 #define HH_V20_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 #include "../picoquant.h"
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -245,5 +246,7 @@ void hh_v20_tttr_header_printf(FILE *stream_out,
 		hh_v20_tttr_header_t *tttr_header);
 void hh_v20_tttr_header_fwrite(FILE *stream_out, 
 		hh_v20_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/interactive.h
+++ b/src/interactive.h
@@ -32,6 +32,8 @@
 #ifndef INTERACTIVE_H_
 #define INTERACTIVE_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "types.h"
@@ -47,5 +49,7 @@ typedef void (*pq_interactive_bin_print_t)(FILE *, pq_interactive_bin_t *);
 
 void pq_interactive_bin_printf(FILE *out_stream, pq_interactive_bin_t *bin);
 void pq_interactive_bin_fwrite(FILE *out_stream, pq_interactive_bin_t *bin);
+
+#pragma pack(pop)
 
 #endif 

--- a/src/picoharp/ph_v20.h
+++ b/src/picoharp/ph_v20.h
@@ -32,6 +32,8 @@
 #ifndef PH_V20_H_
 #define PH_V20_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "../picoquant.h"
@@ -39,7 +41,6 @@
 
 #define PH_V20_BASE_RESOLUTION 4e-12
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -247,5 +248,7 @@ void ph_v20_tttr_header_printf(FILE *stream_out,
 		ph_v20_tttr_header_t *tttr_header);
 void ph_v20_tttr_header_fwrite(FILE *stream_out,
 		ph_v20_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/picoquant.h
+++ b/src/picoquant.h
@@ -32,6 +32,8 @@
 #ifndef PICOQUANT_H_
 #define PICOQUANT_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "types.h"
@@ -52,5 +54,7 @@ pq_dispatch_t pq_dispatch_get(options_t *options, pq_header_t *pq_header);
 void pq_resolution_print(FILE *out_stream,
 		int curve, float64_t resolution, 
 		options_t *options);
-	
+
+#pragma pack(pop)
+
 #endif

--- a/src/t2.h
+++ b/src/t2.h
@@ -32,6 +32,8 @@
 #ifndef T2_H_
 #define T2_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "types.h"
@@ -51,5 +53,7 @@ int pq_t2_stream(FILE *stream_in, FILE *stream_out, pq_t2_decode_t decode,
 int pq_t2_next(FILE *stream_in, pq_t2_decode_t decode, tttr_t *tttr, t2_t *t2);
 int pq_t2_fprintf(FILE *stream_out, t2_t *record);
 int pq_t2_fwrite(FILE *stream_out, t2_t *record);
+
+#pragma pack(pop)
 
 #endif

--- a/src/t3.h
+++ b/src/t3.h
@@ -32,6 +32,8 @@
 #ifndef T3_H_
 #define T3_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "types.h"
@@ -55,5 +57,7 @@ int pq_t3_fprintf(FILE *stream_out, t3_t *record);
 int pq_t3_fwrite(FILE *stream_out, t3_t *record);
 
 void pq_t3_to_t2(t3_t *record_in, t2_t *record_out, tttr_t *tttr);
+
+#pragma pack(pop)
 
 #endif

--- a/src/timeharp/th_v20.h
+++ b/src/timeharp/th_v20.h
@@ -32,12 +32,13 @@
 #ifndef TH_V20_H_
 #define TH_V20_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "../picoquant.h"
 #include "../tttr.h"
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -186,5 +187,7 @@ void th_v20_tttr_header_printf(FILE *stream_out,
 		th_v20_tttr_header_t *tttr_header);
 void th_v20_tttr_header_fwrite(FILE *stream_out,
 		th_v20_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/timeharp/th_v30.h
+++ b/src/timeharp/th_v30.h
@@ -32,12 +32,13 @@
 #ifndef TH_V30_H_
 #define TH_V30_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "../picoquant.h"
 #include "../tttr.h"
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -186,5 +187,7 @@ void th_v30_tttr_header_printf(FILE *stream_out,
 		th_v30_tttr_header_t *tttr_header);
 void th_v30_tttr_header_fwrite(FILE *stream_out,
 		th_v30_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/timeharp/th_v50.h
+++ b/src/timeharp/th_v50.h
@@ -32,12 +32,13 @@
 #ifndef TH_V50_H_
 #define TH_V50_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "../picoquant.h"
 #include "../tttr.h"
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -188,5 +189,7 @@ void th_v50_tttr_header_printf(FILE *stream_out,
 		th_v50_tttr_header_t *tttr_header);
 void th_v50_tttr_header_fwrite(FILE *stream_out,
 		th_v50_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/timeharp/th_v60.h
+++ b/src/timeharp/th_v60.h
@@ -32,12 +32,13 @@
 #ifndef TH_V60_H_
 #define TH_V60_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "../picoquant.h"
 #include "../tttr.h"
 
-#pragma pack(2)
 typedef struct {
 	int32_t MapTo;
 	int32_t Show;
@@ -192,5 +193,7 @@ void th_v60_tttr_header_printf(FILE *stream_out,
 		th_v60_tttr_header_t *tttr_header);
 void th_v60_tttr_header_fwrite(FILE *stream_out,
 		th_v60_tttr_header_t *tttr_header);
+
+#pragma pack(pop)
 
 #endif

--- a/src/tttr.h
+++ b/src/tttr.h
@@ -32,6 +32,8 @@
 #ifndef TTTR_H_
 #define TTTR_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 
 #include "types.h"
@@ -47,5 +49,7 @@ typedef struct {
 } tttr_t;
 
 void tttr_marker_print(FILE *stream_out, uint64_t marker);
+
+#pragma pack(pop)
 
 #endif

--- a/src/unified.c
+++ b/src/unified.c
@@ -32,6 +32,7 @@ POSSIBILITY OF SUCH DAMAGE.
 #include "unified.h"
 #include "error.h"
 
+#include <inttypes.h>
 #include <math.h>
 #include <stdlib.h>
 #include <string.h>
@@ -187,11 +188,11 @@ int pu_tags_read(FILE *stream_in, FILE *stream_out, pu_header_t *pu_header, opti
 					pu_options->number_of_records = tag.value;
 				}
 
-				TAG_PRINT(fprintf(stream_out, "%ld", (uint64_t)tag.value))
+				TAG_PRINT(fprintf(stream_out, "%" PRId64, (int64_t)tag.value))
 				break;
 			case PU_TAG_BitSet64:
 			case PU_TAG_Color8:  // just print both BitSet64 and Color8 for now
-				TAG_PRINT(fprintf(stream_out, "0x%lx", (uint64_t)tag.value))
+				TAG_PRINT(fprintf(stream_out, "0x%" PRIu64 "x", (uint64_t)tag.value))
 				break;
 			case PU_TAG_Float8:
 				memcpy((char *)&value_float, (char *)(&tag.value), sizeof(float64_t));
@@ -203,7 +204,7 @@ int pu_tags_read(FILE *stream_in, FILE *stream_out, pu_header_t *pu_header, opti
 				TAG_PRINT(fprintf(stream_out, "%E", value_float))
 				break;
 			case PU_TAG_TDateTime:
-				TAG_PRINT(fprintf(stream_out, "0x%016lx", tag.value))
+				TAG_PRINT(fprintf(stream_out, "0x%016" PRIx64, tag.value))
 				break;
 			case PU_TAG_Float8Array:
 				buffer_float64 = (float64_t *)malloc(tag.value*sizeof(float64_t));

--- a/src/unified.c
+++ b/src/unified.c
@@ -131,7 +131,7 @@ int pu_dispatch(FILE *stream_in, FILE *stream_out, pu_header_t *pu_header, optio
 				case PU_RECORD_TH_260_NT2:
 				case PU_RECORD_TH_260_PT3:
 				case PU_RECORD_TH_260_PT2:
-					error("This mode is not implemented due to a lack of test data. Please open a pull request at https://github.com/tsbischof/picoquant-sample-data and we will get to work on the decoder.\n")
+					error("This mode is not implemented due to a lack of test data. Please open a pull request at https://github.com/tsbischof/picoquant-sample-data and we will get to work on the decoder.\n");
 					NOT_IMPLEMENTED;
 				default:
 					error("Unknown record type:  0x%08lx\n", pu_options.record_type);

--- a/src/unified.h
+++ b/src/unified.h
@@ -1,6 +1,8 @@
 #ifndef UNIFIED_H_
 #define UNIFIED_H_
 
+#pragma pack(push, 2)
+
 #include <stdio.h>
 #include "header.h"
 #include "options.h"
@@ -46,5 +48,7 @@ typedef struct pu_options_t {
 
 int pu_dispatch(FILE *stream_in, FILE *stream_out, pu_header_t *pu_header, options_t *options);
 int pu_tags_read(FILE *stream_in, FILE *stream_out, pu_header_t *pu_header, options_t *options, pu_options_t *pu_options);
+
+#pragma pack(pop)
 
 #endif

--- a/test.py
+++ b/test.py
@@ -26,8 +26,9 @@ def run(file_in, *args):
         err = stderr.decode()
     return raw
 
+
 warnings.warn(
-            """The current decoded test data were generated with an unknown version of picoquant. 
+    """The current decoded test data were generated with an unknown version of picoquant. 
 If you observe a failure for one of these data files, check whether your current version 
 of picoquant appear to be generating reasonable data -- it might be the case that there was
 a bug in an old version.
@@ -35,7 +36,8 @@ If you run into an issue, contact the maintainer, who can work with you to verif
 the output is correct.
 
 This warning will be removed once all existing data has a verified decoded value"""
-        )
+)
+
 
 class PicoquantTestCase(unittest.TestCase):
     def test_data(self):
@@ -59,18 +61,31 @@ class PicoquantTestCase(unittest.TestCase):
             with self.subTest(binary_file_path=binary_file_path):
                 header_path = binary_file_path + ".header"
                 if not os.path.exists(header_path):
-                    self.skipTest("decoded .header file not available for {binary_file_path=}")
+                    self.skipTest(
+                        f"decoded .header file not available for {binary_file_path=}"
+                    )
 
                 content = run(binary_file_path, "--header-only")
                 with open(header_path, "rb") as f:
                     reference = f.read().decode()
 
-                if content != reference:
-                    print(len(content), len(reference))
-                    print(content.encode())
-                    print(reference.encode())
+                self.assertTrue(content == reference)
+
+    def test_resolution(self):
+        for binary_file_path in binary_file_paths():
+            with self.subTest(binary_file_path=binary_file_path):
+                resolution_path = binary_file_path + ".resolution"
+                if not os.path.exists(resolution_path):
+                    self.skipTest(
+                        "decoded .resolution file not available for {binary_file_path=}"
+                    )
+
+                content = run(binary_file_path, "--resolution-only")
+                with open(resolution_path, "rb") as f:
+                    reference = f.read().decode()
 
                 self.assertTrue(content == reference)
+
 
 if __name__ == "__main__":
     unittest.main()

--- a/test.py
+++ b/test.py
@@ -38,7 +38,7 @@ This warning will be removed once all existing data has a verified decoded value
             with self.subTest(binary_file_path=binary_file_path):
                 data_path = binary_file_path + ".data"
                 if not os.path.exists(data_path):
-                    continue
+                    self.skipTest(r"decoded .data file not available for {binary_file_path=}")
 
                 content = run(binary_file_path).strip().split("\n")
                 with open(binary_file_path + ".data") as f:

--- a/test.py
+++ b/test.py
@@ -1,0 +1,51 @@
+#!/usr/bin/env python3
+
+import os
+import re
+import subprocess
+import unittest
+import warnings
+
+picoquant = "./src/picoquant"
+binary_file_pattern = re.compile("v.+\.([hpt]hd|t3r|[ph]t[23]|ptu)$")
+
+def binary_file_paths(sample_data_dir="sample_data"):
+    for root, dirs, basenames in os.walk("sample_data"):
+        for basename in filter(binary_file_pattern.match, basenames):
+            path = os.path.join(root, basename)
+            yield path
+
+def run(file_in, *args):
+    cmd = [picoquant, "--file-in", file_in, *args]
+    with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as p:
+        stdout, stderr = p.communicate()
+        raw = stdout.decode()
+        err = stderr.decode()
+    return raw
+
+class PicoquantDataTestCase(unittest.TestCase):
+    def test_data(self):
+        warnings.warn(
+"""The current decoded test data were generated with an unknown version of picoquant. 
+If you observe a failure for one of these data files, check whether your current version 
+of picoquant appear to be generating reasonable data -- it might be the case that there was
+a bug in an old version.
+If you run into an issue, contact the maintainer, who can work with you to verify whether 
+the output is correct.
+
+This warning will be removed once all existing data has a verified decoded value""")
+        for binary_file_path in binary_file_paths():
+            with self.subTest(binary_file_path=binary_file_path):
+                data_path = binary_file_path + ".data"
+                if not os.path.exists(data_path):
+                    continue
+
+                content = run(binary_file_path).strip().split("\n")
+                with open(binary_file_path + ".data") as f:
+                    reference = f.read().strip().split("\n")
+
+                content = content[:len(reference)]
+                self.assertTrue(content == reference)
+
+if __name__ == "__main__":
+    unittest.main()

--- a/test.py
+++ b/test.py
@@ -1,6 +1,5 @@
 #!/usr/bin/env python3
 
-import difflib
 import os
 import re
 import subprocess
@@ -23,19 +22,22 @@ def run(file_in, *args):
     with subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as p:
         stdout, stderr = p.communicate()
         raw = stdout.decode()
-        err = stderr.decode()
+        stderr.decode()
     return raw
 
 
 warnings.warn(
-    """The current decoded test data were generated with an unknown version of picoquant. 
-If you observe a failure for one of these data files, check whether your current version 
-of picoquant appear to be generating reasonable data -- it might be the case that there was
-a bug in an old version.
-If you run into an issue, contact the maintainer, who can work with you to verify whether 
-the output is correct.
+    """The current decoded test data were generated with an unknown version
+    of picoquant. If you observe a failure for one of these data files, 
+    check whether your current version of picoquant appears to be 
+    generating reasonable data -- it might be the case that there was a bug
+    in an old version. 
 
-This warning will be removed once all existing data has a verified decoded value"""
+    If you run into an issue, contact the maintainer, who can work with 
+    you to verify whether the output is correct.
+
+    This warning will be removed once all existing data has a verified 
+    decoded value"""
 )
 
 

--- a/test.py
+++ b/test.py
@@ -1,5 +1,6 @@
 #!/usr/bin/env python3
 
+import difflib
 import os
 import re
 import subprocess
@@ -9,11 +10,13 @@ import warnings
 picoquant = "./src/picoquant"
 binary_file_pattern = re.compile("v.+\.([hpt]hd|t3r|[ph]t[23]|ptu)$")
 
+
 def binary_file_paths(sample_data_dir="sample_data"):
     for root, dirs, basenames in os.walk("sample_data"):
         for basename in filter(binary_file_pattern.match, basenames):
             path = os.path.join(root, basename)
             yield path
+
 
 def run(file_in, *args):
     cmd = [picoquant, "--file-in", file_in, *args]
@@ -23,28 +26,50 @@ def run(file_in, *args):
         err = stderr.decode()
     return raw
 
-class PicoquantDataTestCase(unittest.TestCase):
-    def test_data(self):
-        warnings.warn(
-"""The current decoded test data were generated with an unknown version of picoquant. 
+warnings.warn(
+            """The current decoded test data were generated with an unknown version of picoquant. 
 If you observe a failure for one of these data files, check whether your current version 
 of picoquant appear to be generating reasonable data -- it might be the case that there was
 a bug in an old version.
 If you run into an issue, contact the maintainer, who can work with you to verify whether 
 the output is correct.
 
-This warning will be removed once all existing data has a verified decoded value""")
+This warning will be removed once all existing data has a verified decoded value"""
+        )
+
+class PicoquantTestCase(unittest.TestCase):
+    def test_data(self):
         for binary_file_path in binary_file_paths():
             with self.subTest(binary_file_path=binary_file_path):
                 data_path = binary_file_path + ".data"
                 if not os.path.exists(data_path):
-                    self.skipTest(r"decoded .data file not available for {binary_file_path=}")
+                    self.skipTest(
+                        r"decoded .data file not available for {binary_file_path=}"
+                    )
 
                 content = run(binary_file_path).strip().split("\n")
-                with open(binary_file_path + ".data") as f:
+                with open(data_path) as f:
                     reference = f.read().strip().split("\n")
 
-                content = content[:len(reference)]
+                content = content[: len(reference)]
+                self.assertTrue(content == reference)
+
+    def test_header(self):
+        for binary_file_path in binary_file_paths():
+            with self.subTest(binary_file_path=binary_file_path):
+                header_path = binary_file_path + ".header"
+                if not os.path.exists(header_path):
+                    self.skipTest("decoded .header file not available for {binary_file_path=}")
+
+                content = run(binary_file_path, "--header-only")
+                with open(header_path, "rb") as f:
+                    reference = f.read().decode()
+
+                if content != reference:
+                    print(len(content), len(reference))
+                    print(content.encode())
+                    print(reference.encode())
+
                 self.assertTrue(content == reference)
 
 if __name__ == "__main__":

